miner/proposal_builder.go

```diff
@@ -2,62 +2,133 @@
-	errNoBeacon       = errors.New("not building proposals: missing beacon")
-	errDuplicateLayer = errors.New("not building proposals: duplicate layer event")
-)
+//go:generate mockgen -typed -package=mocks -destination=./mocks/mocks.go -source=./proposal_builder.go
+
+type conservativeState interface {
+	SelectProposalTXs(types.LayerID, int) []types.TransactionID
+}
+
+type votesEncoder interface {
+	LatestComplete() types.LayerID
+	TallyVotes(context.Context, types.LayerID)
+	EncodeVotes(context.Context, ...tortoise.EncodeVotesOpts) (*types.Opinion, error)
+}
+
+type mesh interface {
+	GetMalfeasanceProof(nodeID types.NodeID) (*types.MalfeasanceProof, error)
+}
+
+type layerClock interface {
+	AwaitLayer(layerID types.LayerID) <-chan struct{}
+	CurrentLayer() types.LayerID
+	LayerToTime(types.LayerID) time.Time
+}
 
 // ProposalBuilder builds Proposals for a miner.
 type ProposalBuilder struct {
 	logger log.Log
 	cfg    config
-	cdb    *datastore.CachedDB
+
+	cdb       *datastore.CachedDB
+	clock     layerClock
+	publisher pubsub.Publisher
+	conState  conservativeState
+	tortoise  votesEncoder
+	syncer    system.SyncStateProvider
+
+	signer  *signing.EdSigner
+	session *session
 
 	startOnce sync.Once
 	ctx       context.Context
 	cancel    context.CancelFunc
 	eg        errgroup.Group
+}
+
+// session per every signing key for the whole epoch.
+type session struct {
+	epoch     types.EpochID
+	beacon    types.Beacon
+	atx       types.ATXID
+	atxWeight uint64
+	ref       types.BallotID
+	prev      types.LayerID
+	nonce     types.VRFPostIndex
+	active    struct {
+		set    types.ATXIDList
+		weight uint64
+	}
+	eligibilities struct {
+		proofs map[types.LayerID][]types.VotingEligibility
+		slots  uint32
+	}
+}
 
-	clock          layerClock
-	publisher      pubsub.Publisher
-	signer         *signing.EdSigner
-	nonceFetcher   nonceFetcher
-	conState       conservativeState
-	tortoise       votesEncoder
-	proposalOracle proposalOracle
-	beaconProvider system.BeaconGetter
-	syncer         system.SyncStateProvider
+func (s *session) MarshalLogObject(encoder log.ObjectEncoder) error {
+	encoder.AddUint32("epoch", s.epoch.Uint32())
+	encoder.AddString("beacon", s.beacon.String())
+	encoder.AddString("atx", s.atx.ShortString())
+	encoder.AddUint64("weight", s.atxWeight)
+	if s.ref != types.EmptyBallotID {
+		encoder.AddString("ref", s.ref.String())
+	}
+	encoder.AddUint32("prev", s.prev.Uint32())
+	if s.eligibilities.proofs != nil {
+		encoder.AddUint32("slots", s.eligibilities.slots)
+		encoder.AddInt("eligible", len(s.eligibilities.proofs))
+		encoder.AddArray("eligible by layer", log.ArrayMarshalerFunc(func(encoder log.ArrayEncoder) error {
+			// Sort the layer map to log the layer data in order
+			keys := make([]types.LayerID, 0, len(s.eligibilities.proofs))
+			for k := range s.eligibilities.proofs {
+				keys = append(keys, k)
+			}
+			sort.Slice(keys, func(i, j int) bool {
+				return keys[i].Before(keys[j])
+			})
```

Comment:
How about:
```suggestion
			keys := maps.Keys(s.eligibilities.proofs)
			slices.Sort(keys)
```
[slices.Sort](https://pkg.go.dev/golang.org/x/exp@v0.0.0-20230905200255-921286631fa9/slices#Sort) sorts in ascending order.